// @module-federation/bridge-react/dist/router-v6.es.js
import React__default, { useContext } from "react";
import * as ReactRouterDom$1 from "react-router-dom/dist/index.js";
import { R as RouterContext, L as LoggerInstance } from "./index-BlBMQSoq.js";
export * from "react-router-dom/dist/index.js";
function WraperRouter(props) {
  const { basename, ...propsRes } = props;
  const routerContextProps = useContext(RouterContext) || {};
  LoggerInstance.debug(`WraperRouter info >>>`, {
    ...routerContextProps,
    routerContextProps,
    WraperRouterProps: props
  });
  if (routerContextProps == null ? void 0 : routerContextProps.memoryRoute) {
    return /* @__PURE__ */ React__default.createElement(
      ReactRouterDom$1.MemoryRouter,
      {
        ...props,
        initialEntries: [routerContextProps == null ? void 0 : routerContextProps.memoryRoute.entryPath]
      }
    );
  }
  return /* @__PURE__ */ React__default.createElement(
    ReactRouterDom$1.BrowserRouter,
    {
      ...propsRes,
      basename: (routerContextProps == null ? void 0 : routerContextProps.basename) || basename
    }
  );
}
function WraperRouterProvider(props) {
  const { router, ...propsRes } = props;
  const routerContextProps = useContext(RouterContext) || {};
  const routers = router.routes;
  LoggerInstance.debug(`WraperRouterProvider info >>>`, {
    ...routerContextProps,
    routerContextProps,
    WraperRouterProviderProps: props,
    router
  });
  const RouterProvider = ReactRouterDom$1["RouterProvider"];
  const createMemoryRouter = ReactRouterDom$1["createMemoryRouter"];
  const createBrowserRouter = ReactRouterDom$1["createBrowserRouter"];
  if (routerContextProps.memoryRoute) {
    const MemeoryRouterInstance = createMemoryRouter(routers, {
      initialEntries: [routerContextProps == null ? void 0 : routerContextProps.memoryRoute.entryPath]
    });
    return /* @__PURE__ */ React__default.createElement(RouterProvider, { router: MemeoryRouterInstance });
  } else {
    const BrowserRouterInstance = createBrowserRouter(routers, {
      // In host app, the routerContextProps is {}, so we should use router.basename as fallback
      basename: routerContextProps.basename || router.basename,
      future: router.future,
      window: router.window
    });
    return /* @__PURE__ */ React__default.createElement(RouterProvider, { ...propsRes, router: BrowserRouterInstance });
  }
}
export {
  WraperRouter as BrowserRouter,
  WraperRouterProvider as RouterProvider
};
